sympy/physics/mechanics/joint.py

```diff
@@ -0,0 +1,351 @@
+from sympy.physics.vector.frame import ReferenceFrame
+from sympy.physics.mechanics.body import Body
+from sympy.physics.vector import Vector, dynamicsymbols, cross
+from abc import ABC, abstractmethod
+
+__all__ = ['Joint', 'PinJoint']
+
+
+class Joint(ABC):
+    """Abstract Base class for all specific joints.
+
+    Explanation
+    ===========
+
+    A joint subtracts degrees of freedom from a body.
+    This is the base class for all specific joints and holds all common methods
+    acting as an interface for all joints. Custom joint can be created by
+    inheriting Joint class and defining all abstract functions.
+
+    Parameters
+    ==========
+
+    name : string
+        The Joint's name which makes it unique.
+    parent : Body
+        The parent body of joint.
+    child : Body
+        The child body of joint.
+    coordinates: List of dynamicsymbols/dynamicsymbol, optional
+        Coordinates of joint.
+    speeds : List of dynamicsymbols/dynamicsymbol, optional
+        Speeds of joint.
+    parent_joint_pos : Vector, optional
+        Defines the joint's point where the parent will be connected to child.
+        Default value is masscenter of Parent Body.
+    child_joint_pos : Vector, optional
+        Defines the joint's point where the child will be connected to parent.
+        Default value is masscenter of Child Body.
+    parent_axis : Vector, optional
+        Axis of parent frame which would be aligned with child's
+        axis. Default is x axis in parent's frame.
+    child_axis : Vector, optional
+        Axis of child frame which would be aligned with parent's
+        axis. Default is x axis in child's frame.
```

Comment:
These four vectors should all be constant with respect to time. A check is needed to ensure that for each vector.

Reply:
Done